presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java

```diff
@@ -294,6 +302,45 @@ private Estimate calculateNullsFractionForPartitioningKey(HiveColumnHandle parti
         return new Estimate(estimatedNullsCount / estimatedTotalRowsCount);
     }
 
+    private Estimate calculateDataSizeForPartitioningKey(HiveColumnHandle partitionColumn, List<HivePartition> queriedPartitions, Estimate rowCount)
+    {
+        String baseType = partitionColumn.getTypeSignature().getBase();
+        if (!VARCHAR.equals(baseType) && !CHAR.equalsIgnoreCase(baseType)) {
+            return Estimate.unknownValue();
+        }
+
+        OptionalDouble averageDataSize = queriedPartitions.stream()
+                .map(HivePartition::getKeys)
+                .map(keys -> keys.get(partitionColumn))
+                .map(NullableValue::getValue)
+                .mapToDouble(object -> object != null ? ((Slice) object).length() : 0)
+                .average();
+
+        if (averageDataSize.isPresent() && !rowCount.isValueUnknown()) {
+            return new Estimate(averageDataSize.getAsDouble() * rowCount.getValue());
+        }
+        return Estimate.unknownValue();
+    }
+
+    private Estimate calculateDataSize(Map<String, PartitionStatistics> statisticsSample, String columnName, Estimate rowCount)
+    {
+        OptionalDouble averageDataSize = statisticsSample.values().stream()
+                .map(PartitionStatistics::getColumnStatistics)
+                .map(columnStatistics -> columnStatistics.get(columnName))
+                .filter(Objects::nonNull)
+                .map(HiveColumnStatistics::getAverageColumnLength)
+                .filter(OptionalDouble::isPresent)
+                .mapToDouble(OptionalDouble::getAsDouble)
+                .average();
+
+        if (averageDataSize.isPresent() && !rowCount.isValueUnknown()) {
+            return new Estimate(averageDataSize.getAsDouble() * rowCount.getValue());
```

Comment:
does averageDataSize the average over all rows or only non-null rows?  if non-null rows, you need to account for that.

Reply:
@sopel39's internal PR to update this says this is for non-null rows only, will update